index.js

```diff
@@ -1,300 +1,147 @@
-      const b64part = boxedStr.substring(0, boxedStr.length - '.box'.length)
-      const d = Buffer.from(b64part, 'base64')
-      return Buffer.concat([BOX1_TF, d])
-    } else if (boxedStr.endsWith('.box2')) {
-      const b64part = boxedStr.substring(0, boxedStr.length - '.box2'.length)
-      const d = Buffer.from(b64part, 'base64')
-      return Buffer.concat([BOX2_TF, d])
-    } else throw new Error('Unknown boxed string: ' + boxedStr)
-  },
+function encode(input) {
+  /* cases we don't encode */
+  if (input === undefined || Buffer.isBuffer(input) || Number.isInteger(input))
+    return input
 
-  signature(sig) {
-    const b64part = sig.substring(0, sig.length - '.sig.ed25519'.length)
-    const d = Buffer.from(b64part, 'base64')
-    return Buffer.concat([SIGNATURE_TF, d])
-  },
+  if (typeof input === 'string') {
+    /* looks for classic sigil/suffix matches */
+    const { type, format } = findClassicTypeFormat(input, TYPES)
+    if (type) {
+      if (format) return encoder.classic(input, type, format)
+      else throw new Error(`Unknown ${type.type} format`)
+    }
 
-  string(str) {
-    const d = Buffer.from(str, 'utf8')
-    return Buffer.concat([STRING_TF, d])
-  },
+    /* fallback */
+    return encoder.string(input)
+  }
 
-  boolean(bool) {
-    const d = bool ? BOOL_TRUE : BOOL_FALSE
-    return Buffer.concat([BOOL_TF, d])
-  },
-}
+  if (typeof input === 'boolean') return encoder.boolean(input)
+  if (input === null) return NIL.TFD
 
-function encode(input) {
+  /* recursions */
   if (Array.isArray(input)) {
     return input.map((x) => {
       const y = encode(x)
-      if (y === undefined) return NIL_TFD
+      if (y === undefined) return NIL.TFD
       else return y
     })
   }
-  if (input === undefined) {
-    return undefined
-  } else if (input === null) {
-    return NIL_TFD
-  } else if (typeof input === 'object' && !Buffer.isBuffer(input)) {
+  if (typeof input === 'object') {
+    // we know it's not: Buffer,null,Array
     const output = {}
-    for (let key in input) {
-      const x = input[key]
-      const y = encode(x)
+    for (const key in input) {
+      const y = encode(input[key])
       if (y !== undefined) output[key] = y
     }
     return output
-  } else if (typeof input === 'string') {
-    if (isFeedType(input)) return encoder.feed(input)
-    else if (isMsgType(input)) return encoder.message(input)
-    else if (isBlobType(input)) return encoder.blob(input)
-    else if (input.endsWith('.sig.ed25519')) return encoder.signature(input)
-    else if (input.match(/\.box\d*$/)) return encoder.box(input)
-    else return encoder.string(input)
-  } else if (typeof input === 'boolean') {
-    return encoder.boolean(input)
-  } else {
-    if (!Number.isInteger(input) && !Buffer.isBuffer(input))
-      console.warn('not encoding unknown value', input)
-    // FIXME: more checks, including floats!
-    return input
   }
+
+  throw new Error('cannot encoding, type is not defined')
```

Comment:
In all the errors, we have been trying to provide the faulty data, so the error is useful in a debugging context. I suggest changing this one to tell what the faulty data is.

Reply:
agree, will change